python/ray/data/datasource/databricks_uc_datasource.py

```diff
@@ -0,0 +1,187 @@
+import json
+import time
+
+import logging
+import requests
+import os
+from urllib.parse import urljoin, urlencode
+
+import pyarrow
+import requests
+
+import math
+from contextlib import contextmanager
+from typing import Any, Callable, Iterable, Iterator, List, Optional
+
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+from ray.util.annotations import PublicAPI
+
+
+logger = logging.getLogger(__name__)
+
+
+class _DatabricksUCReader(Reader):
+    def __init__(
+        self,
+        host: str,
+        token: str,
+        warehouse_id: str,
+        catalog: str,
+        schema: str,
+        query: str,
+    ):
+        self.host = host
+        self.token = token
+        self.warehouse_id = warehouse_id
+        self.catalog = catalog
+        self.schema = schema
+        self.query = query
+
+        url_base = f"https://{self.host}/api/2.0/sql/statements/"
+
+        payload = json.dumps({
+            "statement": self.query,
+            "warehouse_id": self.warehouse_id,
+            "wait_timeout": "0s",
+            "disposition": "EXTERNAL_LINKS",
+            "format": "ARROW_STREAM",
+            "catalog": self.catalog,
+            "schema": self.schema,
+        })
+
+        req_auth = ('token', self.token)
+        req_headers = {'Content-Type': 'application/json'}
+
+        response = requests.post(
+            url_base,
+            auth=('token', self.token),
+            headers={'Content-Type': 'application/json'},
+            data=payload
+        )
+        statement_id = response.json()["statement_id"]
+
+        state = response.json()["status"]["state"]
+
+        logger.info(f"Waiting for query {query} exeuction result.")
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
+                response = requests.get(
+                    urljoin(url_base, statement_id) + "/",
+                    auth=req_auth,
+                    headers=req_headers,
+                )
+                response.raise_for_status()
+                state = response.json()["status"]["state"]
```

Comment:
no timeout?

Reply:
I see, we don't know how long `query` takes.

Reply:
We allow user to press Ctrl + C to cancel it, this should fulfil most cases

Reply:
How long query takes is hard to estimate, it depends on sql workloads, query queues, warehouse resources, etc.